src/VkLayer_GF_amber_scoop/src/buffer_copy.cc

```diff
@@ -0,0 +1,208 @@
+
+  // Create command buffer
+  VkCommandBuffer command_buffer = nullptr;
+  VkCommandBufferAllocateInfo command_buffer_allocate_info = {};
+  command_buffer_allocate_info.sType =
+      VK_STRUCTURE_TYPE_COMMAND_BUFFER_ALLOCATE_INFO;
+  command_buffer_allocate_info.commandPool = command_pool;
+  command_buffer_allocate_info.level = VK_COMMAND_BUFFER_LEVEL_PRIMARY;
+  command_buffer_allocate_info.commandBufferCount = 1;
+  RequireSuccess(device_data_->vkAllocateCommandBuffers(
+                     device, &command_buffer_allocate_info, &command_buffer),
+                 "Failed to allocate command buffers.");
+
+  // Record a command buffer for the copy operation.
+  VkCommandBufferBeginInfo begin_info = {};
+  begin_info.sType = VK_STRUCTURE_TYPE_COMMAND_BUFFER_BEGIN_INFO;
+
+  RequireSuccess(
+      device_data_->vkBeginCommandBuffer(command_buffer, &begin_info),
+      "Failed to begin command buffer.");
+
+  // Copy all global and buffer memory barriers. Set
+  // VK_ACCESS_TRANSFER_READ_BIT for copied (buffer)memory barriers, so we can
+  // execute a copy operation.
+  for (const auto& pipeline_barrier : pipeline_barriers) {
+    auto buffer_memory_barriers = std::vector<VkBufferMemoryBarrier>(
+        pipeline_barrier->GetBufferMemoryBarriers());
+    auto memory_barriers =
+        std::vector<VkMemoryBarrier>(pipeline_barrier->GetMemoryBarriers());
+
+    // Set dstAccessMask(s) to VK_ACCESS_TRANSFER_READ_BIT
+    for (auto& buffer_memory_barrier : buffer_memory_barriers) {
+      buffer_memory_barrier.dstAccessMask = VK_ACCESS_TRANSFER_READ_BIT;
+    }
+    for (auto& memory_barrier : memory_barriers) {
+      memory_barrier.dstAccessMask = VK_ACCESS_TRANSFER_READ_BIT;
+    }
+
+    // clang-format off
+    device_data_->vkCmdPipelineBarrier(
+        command_buffer,
+        pipeline_barrier->GetSrcStageMask(),
+        VK_PIPELINE_STAGE_TRANSFER_BIT,
+        0,
+        static_cast<uint32_t>(memory_barriers.size()),
+        memory_barriers.data(),
+        static_cast<uint32_t>(buffer_memory_barriers.size()),
+        buffer_memory_barriers.data(),
+        0,
+        nullptr);
+    // clang-format on
+  }
+
+  // Copy buffer
+  VkBufferCopy copy_region = {0, 0, buffer_size};
+  device_data_->vkCmdCopyBuffer(command_buffer, buffer, buffer_copy_, 1,
+                                &copy_region);
+
+  // Memory barrier for making sure we can safely read the copied data from
+  // the host.
+  VkMemoryBarrier host_copy_barrier = {};
+  host_copy_barrier.sType = VK_STRUCTURE_TYPE_MEMORY_BARRIER;
+  host_copy_barrier.srcAccessMask = VK_ACCESS_TRANSFER_WRITE_BIT;
+  host_copy_barrier.dstAccessMask = VK_ACCESS_HOST_READ_BIT;
+  // clang-format off
+  device_data_->vkCmdPipelineBarrier(
+      command_buffer,
+      VK_PIPELINE_STAGE_TRANSFER_BIT,
+      VK_PIPELINE_STAGE_HOST_BIT,
+      0,
+      1,
+      &host_copy_barrier,
+      0,
+      nullptr,
+      0,
+      nullptr);
```

Comment:
I don't think it's actually ideal to use a buffer memory barrier. At least Khronos synchronization examples guides to use per-resource barriers only for queue ownership transfers (and image layout transitions in case of image buffer barrier) https://github.com/KhronosGroup/Vulkan-Docs/wiki/Synchronization-Examples#three-dispatches-first-dispatch-writes-to-one-storage-buffer-second-dispatch-writes-to-a-different-storage-buffer-third-dispatch-reads-both

Reply:
Ah interesting.